### PR TITLE
Fix DistributedSampler mem usage on large datasets

### LINKED ISSUE #45427: Memory Consumption of DistributedSampler too large when deal with huge datasets
## To Reproduce

1. create a dataset with large __len__
1. use DistributedSampler as its sampler
1. it could be find that `.tolist()` operation would cause many times of memory than original torch.Tensor object

## Expected behavior

We can delete `to_list` operation and write a simply iterator to prevent huge memory consumption:

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

cc @SsnL @VitalyFedyunin

### PR DESCRIPTION
**Pull request by amccollum**
The current implementation of DistributedSampler generates a python list to hold all of the indices, and then returns a slice of this list for the given rank (creating a partial copy of the list). When the underlying dataset is large, both of these choices waste a large amount of memory. It is much more efficient to create a tensor to hold the indices, and then index into that tensor instead of creating slices.

In the case of a sampler with `shuffle=False`, it would be possible to avoid creating the `indices` tensor entirely (since the index will always match the value), but I have opted instead here to keep the implementation as similar to the existing version as possible. One possible benefit of this approach is that memory usage will not significantly change based on changing this parameter. Still, it might be better to simply return the indices directly without the underlying array.

Additionally, the logic around calculating the number of samples is unnecessarily complex. When dropping the last batch, this can be a simple floor division.

In a simple test script which creates a sampler for a dataset with a 100,000,000 items, memory usage is reduced 98% compared to the existing implementation.

Fixes #45427

### CODE REVIEW

File: torch/utils/data/distributed.py [hunk 1]
```diff
@@ -76,49 +73,37 @@ def __init__(self, dataset: Dataset, num_replicas: Optional[int] = None,
...
         self.total_size = self.num_samples * self.num_replicas
         self.shuffle = shuffle
         self.seed = seed
 
-    def __iter__(self) -> Iterator[T_co]:
+    def __iter__(self) -> Iterator[int]:
         if self.shuffle:
             # deterministically shuffle based on epoch and seed
             g = torch.Generator()
             g.manual_seed(self.seed + self.epoch)
```
Comment (amccollum) [inline torch/utils/data/distributed.py:88]: Rather than having a separate method, `set_epoch`, to increment the random seed on each epoch, it would be easier (for the user) just to increment `self.seed` here in the iterator. Presumably the sampler isn't being iterated over a variable number of times, so the result should be the same without having the overhead of having to remember to increment the epoch. It's hard to see a situation where you _wouldn't_ want the samples to be reshuffled on each pass.

**Comment by amccollum** [comment]
Is there something I could do to make this pull request more easily merged? I think the outrageous memory usage of DistributedSampler on large datasets is pretty surprising, and probably catches a fair number of folks off guard.

File: torch/utils/data/distributed.py [hunk 2]
```diff
@@ -1,15 +1,12 @@
...
 import torch
 from . import Sampler, Dataset
 import torch.distributed as dist
 
 
-T_co = TypeVar('T_co', covariant=True)
-
-
-class DistributedSampler(Sampler[T_co]):
+class DistributedSampler(Sampler[int]):
```
Comment (rohan-varma) [inline torch/utils/data/distributed.py:9]: Any reason why the typing here is changed?

**Review by rohan-varma — COMMENTED** [review]
Thanks for this awesome change and sorry for the late review! In general, this looks great and the memory savings is quite impressive.

Could you paste your script that you used to calculate the memory savings so that others can benchmark this change as well? Also, would be great if you can rebase the PR and then it should be good to land. Thanks!

**Comment by amccollum** [comment]
Sorry, accidentally bloated the pull request when rebasing.

File: torch/utils/data/distributed.py (hunk 2 above)
Comment (amccollum) [reply torch/utils/data/distributed.py:9]: When switching from returning the list iterator to creating a generator, it makes the type more specific.

**Comment by amccollum** [comment]
Here is a quick snippet to print memory usage of `DistributedSampler`:

```
import os, psutil, torch
sampler = torch.utils.data.distributed.DistributedSampler(range(100_000_000), rank=0, num_replicas=8, shuffle=False)
it = iter(sampler)
next(it)
print(psutil.Process(os.getpid()).memory_info().rss)
```

If you run this before/after the change, you will see a very large reduction in memory usage.

**Review by rohan-varma — APPROVED** [review]
Thanks, LGTM! Will land once all the CI clears

**Comment by rohan-varma** [comment]
Just a quick update on this PR - we are currently running this on some benchmark workloads that use `DistributedSampler` to understand the memory savings here on real-world workloads. After we've run those, this PR should be good to land
